GUIDE.md

```diff
@@ -0,0 +1,259 @@
+working with Git and with GitHub, and I will try to describe the ways that
+we encourage as part of Xdev.
+
+### Use Forks+Branches instead of just Branches.
+
+Whenever development happens, we strongly encourage you to do everything
+in a _fork_, instead of directly pushing changes into the main repository.
+We still encourage the use of _descriptive branch names_ in your fork,
+describing the feature/fix that is being implemented, so that GitHub
+history graphs can be more descriptive than just "master was merged into
+master".  Even if you have rights to directly push into a repository,
+we recommend doing everything through forks and pull requests.
+
+Also, keep in might that the Git best practice of "commit early and
+commit often" also, in a way, applies to forks+branches.  Try to keep each
+fork+branch limited to changes that implement _only one feature_; Do not
+implement multiple features in one fork+branch.  This can seem like an
+inconvenient rule, but it helps others when they look at the history
+later.
+
+### Follow Git best practices.
+
+Whenever possible, we also recommend using standard Git best practices,
+such as:
+
+- _Do_ write descriptive commit messages,
+- _Do_ commit early and commit often,
+- _Don't_ change the published history of a repository.
+
+There are other git best practices that many other teams suggest (and
+sometimes demand), but, for now, we will not demand them.  These include
+the ["squash and merge"](https://blog.pairworking.com/why-you-should-care-about-squash-and-merge-in-git-675856bf66b0)
+rule and the ["no merge commits"](https://shinglyu.com/web/2018/03/25/merge-pull-requests-without-merge-commits.html)
+rule.  If there is demand from the team that we _do_ implement these
+rules, then we can discuss them and how best to implement them.
+
+We also recommend the use of [Git pre-commit hooks](https://ljvmiranda921.github.io/notebook/2018/06/21/precommits-using-black-and-flake8/)
+for formatting code whenever you make new changes.
+
+### Clearly define (and use) project Conda environments.
+
+Whenever working on a project, whether it be software development or
+testing or helping scientists translate code into Python or developing
+educational material and documentation, it is important to do your work
+within a virtual environment that safely provides all of the software
+dependencies you need for your work.  We recommend using Miniconda to
+create and manage these environments.
+
+It can be useful to _share_ these environments with other collaborators
+at times.  If all you need is an environment for testing a specific
+package, we usually keep a Conda YAML environment file in the CI
+directory for the package of interest ([see below](#use-ci-for-as-much-as-possible.)).
+However, if you are doing a project that requires multiple packages that
+are not dependencies of each other, then I advise learning how to easily
+take a "snapshot" of your environment via `conda`:
+
+```bash
```

Comment:
In the upcoming release of conda, they've added `--from-history` argument which allows the user to export their conda environment/requirements but don't list all dependencies but only those they directly depend on / have installed explicitly:

```bash
conda env export --from-history > environment.yml
```

Reply:
I noticed that.  I thought I would leave it out of the GUIDE document, though, because I think it is probably less problematic to explicitly declare all dependencies in the environment (with their version numbers).  Do you agree?

Reply:
...that is, when sharing your environment (i.e., when you want reproducibility).  If you want to specify an environment for "general use" in the future, then I think the `--from-history` argument is very helpful.

Reply:
> I thought I would leave it out of the GUIDE document, though, because I think it is probably less problematic to explicitly declare all dependencies in the environment (with their version numbers)

I concur